flux-refineck/src/constraint_gen.rs

```diff
@@ -539,6 +551,150 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
+
+        f_ins
+            .into_iter()
+            .map(|(p, p_in_ty)| {
+                let oblig_def_id = *f_did.get(&p).unwrap();
+                let p_out_ty = f_out.get(&p).unwrap();
+                let p_rust_fn_sig = self.genv.tcx.mk_fn_sig(
+                    p_in_ty.tuple_fields(),
+                    *p_out_ty,
+                    false,
+                    rustc_hir::Unsafety::Normal,
+                    Abi::Rust,
+                );
+                self.mk_oblig(bound_vars, substs, oblig_def_id, &generics, p_rust_fn_sig)
+            })
+            .collect_vec()
+    }
+
+    fn gather_f_trait_info(
+        &mut self,
+        did: DefId,
+        rust_fn_sig: &rustc_middle::ty::PolyFnSig,
+        actuals: &[flux_middle::intern::Interned<rty::TyS>],
+    ) -> (
+        FxHashMap<ParamTy, rustc_middle::ty::Ty<'tcx>>,
+        FxHashMap<ParamTy, rustc_middle::ty::Ty<'tcx>>,
+        FxHashMap<ParamTy, DefId>,
+    ) {
+        let fn_once_id = self.genv.tcx.lang_items().fn_once_trait().unwrap();
+
+        // 1. Find the closure-param-tys i.e. 'F' that implement FnOnce -- ie are supposed to be closures
+        let mut f_ins = FxHashMap::default();
+        let mut f_out = FxHashMap::default();
+        let mut f_did = FxHashMap::default();
+
+        // 2. connect the param-ty and the closure-def
+        let args = rust_fn_sig.skip_binder().inputs();
+        iter::zip(actuals, args).for_each(|(actual, arg)| {
+            if let rustc_middle::ty::Param(p) = arg.kind() &&
+                let rty::TyKind::Indexed(rty::BaseTy::Closure(did),_) = actual.kind() {
+                f_did.insert(*p, *did);
+            }
+        });
+
+        // 3. collect inputs-and-outputs for each closure-def
+        for clause in self.def_id_clauses(did) {
+            match clause {
+                // found an fn-input type
+                Clause::Trait(trait_pred) => {
+                    let trait_ref = trait_pred.trait_ref;
+                    if fn_once_id == trait_ref.def_id &&
+                           let Some(substs) = trait_ref.substs.try_as_type_list() &&
+                           let rustc_middle::ty::Param(p) = substs[0].kind() {
+                            f_ins.insert(*p,substs[1]);
+                        }
+                }
+                // found an fn-output type
+                Clause::Projection(proj_pred) => {
+                    let proj_ty = proj_pred.projection_ty;
+                    // ASSUME: the only trait item of `FnOnce` is `Output`
+                    if self.genv.tcx.trait_of_item(proj_ty.def_id) == Some(fn_once_id) &&
+                           let Some(proj_substs) = proj_ty.substs.try_as_type_list() &&
+                           let rustc_middle::ty::Param(p) = proj_substs[0].kind() &&
+                           let Some(out_ty) = proj_pred.term.ty() {
+                           f_out.insert(*p,out_ty);
+                        }
```

Comment:
`FnOnce::Output` is also a [lang item](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/lang_items/enum.LangItem.html#variant.FnOnceOutput). You can get it with `self.genv.tcx.lang_items().fn_once_output().unwrap()`

Reply:
good catch!